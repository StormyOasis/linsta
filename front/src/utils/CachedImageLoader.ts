import { clear, del, get, set } from 'idb-keyval';
import { blobToBase64 } from "./utils";

export const loadImageCached = async (id: string, url: string): Promise<string> => {
    // First try to get the image from the cache
    const value = await get(id);

    if(value === undefined) {
        // Image not in the cache so have to fetch it
        try {
            const result = await fetch(url);
            if(result.status === 200) {
                const blob = await result.blob();
                const b64:string = await blobToBase64(blob) as string;

                // add the base64 data into the cache and return it
                await set(id, b64);

                return b64;
            }
        } catch(err) {
            console.log(err);
            return url; //default to what was passed in
        }

    } else {
        // Image is in the cache already as a base64 string, use that as is
        return value;
    }

    return url; //default to what was passed in
}

export const setImage = async (id: string, data: string) => {     
    await set(id, data);
}

export const deleteFromCache = async (id: string) => {
    // First try to get the image from the cache
    const value = await get(id);

    if(value === undefined) {
        // key not present in cache so do nothing
        return;
    }

    // delete from the cache
    await del(id);
}

export const clearCache = async () => {
    await clear();
}