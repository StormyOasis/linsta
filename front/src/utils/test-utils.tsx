import React, { ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import { RootState } from "../Components/Redux/redux";
import { authInitialState } from '../Components/Redux/slices/auth.slice';
import { modalInitialState } from '../Components/Redux/slices/modals.slice';
import { profileInitialState } from '../Components/Redux/slices/profile.slice';
import { miscInitialState } from '../Components/Redux/slices/misc.slice';
import { configureStore } from '@reduxjs/toolkit';

import Theme from '../Components/Themes/Theme';

export const mockInitialState = (overrides?: Partial<RootState>): RootState => ({
    auth: { ...authInitialState, ...(overrides?.auth ?? {}) },
    modal: { ...modalInitialState, ...(overrides?.modal ?? {}) },
    profile: { ...profileInitialState, ...(overrides?.profile ?? {}) },
    misc: { ...miscInitialState, ...(overrides?.misc ?? {}) },
});

type ExtendedRenderOptions = {
    store?: ReturnType<typeof configureStore>;
    preloadedState?: any;
    route?: string; // ‚Üê allow overriding the initial route
} & Omit<RenderOptions, "queries">;

const dummyAuthReducer = (state = { user: null, isLoading: false, error: null }, _action: any) => state;
const dummyModalReducer = (state = { isOpen: false }, _action: any) => state;
const dummyProfileReducer = (state = { name: '', email: '' }, _action: any) => state;
const dummyMiscReducer = (state = { someData: null }, _action: any) => state;

const renderWithStore = (
    ui: ReactElement,
    {
        preloadedState,
        store = configureStore({
            reducer: {
                auth: dummyAuthReducer,
                modal: dummyModalReducer,
                profile: dummyProfileReducer,
                misc: dummyMiscReducer,
            },
            preloadedState,
        }),
        route = "/",
        ...renderOptions
    }: ExtendedRenderOptions = {}
) => {
    const Wrapper: React.FC<{ children: React.ReactNode }> = ({ children }) => {
        return (
            <Provider store={store}>
                <MemoryRouter initialEntries={[route]}>
                    <Theme>{children}</Theme>
                </MemoryRouter>
            </Provider>
        );
    };

    return {
        ...render(ui, { wrapper: Wrapper, ...renderOptions }),
        store,
    };
};

export * from "@testing-library/react";
export { renderWithStore };
export type { ExtendedRenderOptions };