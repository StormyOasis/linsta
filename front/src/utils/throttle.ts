import { useRef } from 'react';

const useThrottle = (callback: Function, delay: number) => {
    const lastCall = useRef(0);

    return (...args: any[]) => {
        const now = new Date().getTime();
        if (now - lastCall.current >= delay) {
            lastCall.current = now;
            callback(...args);
        }
    };
};

export default useThrottle;