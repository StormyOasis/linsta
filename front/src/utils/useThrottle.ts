import { useRef, useCallback } from 'react';

/**
 * React hook for throttling a callback function.
 * The callback will only be invoked at most once every `delay` milliseconds.
 *
 * @param callback - The function to throttle.
 * @param delay - The minimum delay (ms) between calls.
 * @returns A throttled version of the callback.
 */
const useThrottle = <T extends (...args: any[]) => void>(callback: T, delay: number) => {
    const lastCall = useRef<number>(0);

    return useCallback(
        (...args: Parameters<T>) => {
            const now = performance.now();
            if (now - lastCall.current >= delay) {
                lastCall.current = now;
                callback(...args);
            }
        },
        [callback, delay]
    );
};

export default useThrottle;