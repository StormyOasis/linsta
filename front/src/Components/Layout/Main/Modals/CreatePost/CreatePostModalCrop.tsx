import React, { useState } from "react";
import styled from "styled-components";
import Cropper, { Area, Point } from 'react-easy-crop';
import { isVideoFileFromType } from "../../../../../utils/utils";
import { ModalSectionWrapper } from "../../../../Common/MultiStepModal";
import Slider from "../../../../Common/Slider";
import { Div, Flex } from "../../../../Common/CombinedStyling";
import MediaSliderButton from "../../../../Common/MediaSliderButton";
import { CropSVG, FourToFiveSVG, ImageSVG, OneToOneSVG, SixteenToNineSVG } from "../../../../../Components/Common/Icon";

const CropContainer = styled(Div)`
    height: ${props => props.theme['sizes'].cropperHeight};
    width: calc(${props => props.theme['sizes'].defaultModalWidth} - 40px);
    max-width: calc(${props => props.theme['sizes'].maxModalWidth} - 40px);
`;

const CropperWrapper = styled(Div)`
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
`;
const CropperContainer = styled(Div)`
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 5px;
`;

const CropperAspectRatioButtonWrapper = styled(Div)`
    z-index: 20;
    position: absolute;
    bottom: 0;
`;

const CropperAspectRatioButton = styled(Div)<{$isopen?: string}>`
    width: 24px;
    height: 24px;
    color: ${props => props.$isopen === "true" ? props.theme['colors'].cropperAspectBkgnd : "white"};
    background-color: ${props => props.$isopen === "true" ? "white" : props.theme['colors'].cropperAspectBkgnd};
    border-radius: 50%;
    padding: 5px; 
    cursor: pointer;
    margin-left: 5px;
    margin-bottom: 10px;

    &:hover {
        color: ${props => props.theme['colors'].borderDefaultColor};
        background-color: ${props => props.theme['colors'].cropperAspectBkgndNoTrans};
    };    
`;

const AspectPopupMenu = styled(Div)`
    z-index: 20;
    position:absolute;
    bottom: 50px;
    left: 10px;
    background-color: ${props => props.theme['colors'].cropperAspectBkgnd};    
    cursor: pointer;
    display: flex;
    flex-direction: column;  
    border-radius: 8px;
    overflow: hidden;
    width: 95px;
`;

const AspectPopupMenuItem = styled(Div)<{selected?: boolean}>`
    padding: 5px;
    border-top: 1px solid ${props => props.theme['colors'].borderDarkColor};    
    color: ${props => props.selected ? "white" : props.theme['colors'].borderDarkColor};
    font-weight: ${props => props.selected ? 700 : 400};
    display: flex;

    &:nth-child(1) {
        border: none;
    };      

    &:hover {
        color: ${props => props.theme['colors'].borderDefaultColor};
        background-color: ${props => props.theme['colors'].cropperAspectBkgndNoTrans};
    };
`;

const AspectPopupMenuItemContent = styled(Div)`
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    width: 100%;
`;

export type CreatePostModalCropProps = {
    files: any[];
    cropData: CropData[];
    setCropData: (cropData: CropData[]) => void;
}

type AspectRatio = {
    value: number;
    text: string;
    icon: any|null;
}

const aspectRatios:AspectRatio[] = [
    { value: 0, text: "Original", icon: {element: <ImageSVG width="22px" height="22px" stroke="none" />}},
    { value: 1 / 1, text: "1:1", icon: {element: <OneToOneSVG width="22px" height="22px"/>}},
    { value: 1 / 2, text: "1:2", icon: null },    
    { value: 4 / 3, text: "4:3", icon: null },
    { value: 4 / 5, text: "4:5", icon: {element: <FourToFiveSVG width="16px" height="22px" />}},
    { value: 16 / 9, text: "16:9", icon: {element: <SixteenToNineSVG width="22px" height="16px"/>}},    
];

export type CropData = {
    crop: Point;
    zoom: number;
    rotation: number;
    aspect: AspectRatio;
    croppedAreaPixels: Area|null;
};

type CropperProps = {
    file: any;
    cropData: CropData;
    hasNext: boolean;
    hasPrev: boolean;
    isVideoFile: boolean;
    onNextFile: () => void;
    onPrevFile: () => void;
    onCropComplete: (
        crop: Point, zoom: number, rotation: number, 
        aspect: AspectRatio, croppedAreaPixels: Area) => void;
};

export const defaultCropData = (count: number):CropData[] => {
   const cropDataList:CropData[] = [];
   for(let i = 0; i < count; i++) {
        cropDataList.push({
            crop: {x: 0, y: 0},
            zoom: 1,
            rotation: 0,
            aspect: aspectRatios[0],
            croppedAreaPixels: null            
        });
   }

   return cropDataList;
}

const CreatePostModalCrop: React.FC<CreatePostModalCropProps> = (props: CreatePostModalCropProps) => {
    const [currentFileIndex, setCurrentFileIndex] = useState(0);
        
    const onNextFile = () => {
        setCurrentFileIndex(currentFileIndex + 1);
    }

    const onPrevFile = () => {
        setCurrentFileIndex(currentFileIndex - 1);
    }

    const onCropComplete = (
        crop: Point, zoom: number, rotation: number, 
        aspect: AspectRatio, croppedAreaPixels: Area) => {

        const newCropData = [...props.cropData];
        newCropData[currentFileIndex] = {
            crop,
            zoom,
            rotation,
            aspect,
            croppedAreaPixels
        }

        props.setCropData(newCropData);
    }

    if(props.files == null || props.files.length === 0) {
        return null;
    }

    const isVideoFile = isVideoFileFromType(props.files[currentFileIndex].type);

    return (
        <>
            {isVideoFile && <Div $fontWeight="700" $color="red" $textAlign="center">Note: Editing video files is currently unsupported</Div>}
            <CreatePostModalCropper 
                isVideoFile={isVideoFile}
                hasNext={currentFileIndex < props.files.length-1}
                hasPrev={currentFileIndex > 0}
                onNextFile={onNextFile}
                onPrevFile={onPrevFile}
                onCropComplete={onCropComplete}
                file={props.files[currentFileIndex]} 
                cropData={props.cropData[currentFileIndex]}
            />
        </>
    );
}

const CreatePostModalCropper: React.FC<CropperProps> = (props: CropperProps) => {
    const [width, setWidth] = useState(0);
    const [height, setHeight] = useState(1);
    const [crop, setCrop] = useState<Point>(props.cropData.crop);
    const [zoom, setZoom] = useState(props.cropData.zoom);    
    const [rotation, setRotation] = useState(props.cropData.rotation);
    const [aspect, setAspect] = useState(props.cropData.aspect);    
    const [_croppedAreaPixels, setCroppedAreaPixels] = useState<Area|null>(props.cropData.croppedAreaPixels);
    const [isAspectMenuOpen, setIsAspectMenuOpen] = useState(false);
    const [isFlaggedForReset, setIsFlaggedForReset] = useState(false);

    const resetStateToProps = () => {
        setCrop(props.cropData.crop);
        setZoom(props.cropData.zoom);
        setRotation(props.cropData.rotation);
        setAspect(props.cropData.aspect);
        setCroppedAreaPixels(props.cropData.croppedAreaPixels);
        setIsAspectMenuOpen(false);
        
        setIsFlaggedForReset(false);
    }

    const getImageDimensions = async (blob: string):Promise<{width: number, height:number}> => {
        return new Promise((resolve, _reject) => {
            const image = new Image();
            image.src = blob;
            image.onload = () => resolve({width: image.width, height: image.height})
        });
    }    

    const onCropComplete = async (_croppedArea: Area, croppedAreaPixels: Area) => {
        let newCrop = crop;
        if(aspect.value === 0) {            
            const {width, height} = await getImageDimensions(props.file.blob);

            newCrop = {x:0, y:0};
            croppedAreaPixels = {
                x: 0,
                y: 0,
                width,
                height,
            };
            setCrop(newCrop);
            setWidth(width);
            setHeight(height);
        }

        setCroppedAreaPixels(croppedAreaPixels);
        
        props.onCropComplete(newCrop, zoom, rotation, aspect, croppedAreaPixels);
    };

    const renderAspectRatioMenu = () => {
        return (
            <AspectPopupMenu>
                {aspectRatios.map((ratio) => {
                    let selected = {selected: (aspect === ratio)};
                
                    return (
                        <AspectPopupMenuItem {...selected} key={ratio.text} onClick={() => setAspect(ratio)}>
                            <AspectPopupMenuItemContent>
                                <Div>{ratio.text}</Div>
                                {ratio.icon && ratio.icon.element}
                            </AspectPopupMenuItemContent>
                        </AspectPopupMenuItem>);
                    })
                }                
            </AspectPopupMenu>
        );
    }    

    let inputProps:any = {};
    if(props.isVideoFile) {
        inputProps = {video: props.file.blob};
    } else {
        inputProps = {image: props.file.blob};
    }

    // If we are flagged for reset we need to resync the local state with the props
    // necessary otherwise we end up with a race condition where we have advanced
    // page number but haven't yet updated the previous page's state. Instead we
    // update the new image's state with now bad data
    if(isFlaggedForReset) {
        resetStateToProps();
    }    

    const aspectRatioValue = (aspect.value !== 0) ? aspect.value : (width / height);

    return (
        <>
            <ModalSectionWrapper>
                <CropContainer>
                    <CropperWrapper>
                        <CropperContainer>
                            <Cropper
                                {...inputProps}
                                crop={crop}
                                rotation={rotation}
                                zoom={zoom}
                                aspect={aspectRatioValue}
                                onCropChange={setCrop}
                                onZoomChange={setZoom}
                                onRotationChange={setRotation}
                                onCropComplete={onCropComplete}                    
                            />
                            {isAspectMenuOpen && renderAspectRatioMenu()}                          
                        </CropperContainer>
                        <CropperAspectRatioButtonWrapper>
                            <CropperAspectRatioButton $isopen={`${isAspectMenuOpen}`} onClick={() => setIsAspectMenuOpen(!isAspectMenuOpen)}>
                                <CropSVG width="24px" height="24px" />
                            </CropperAspectRatioButton>                            
                        </CropperAspectRatioButtonWrapper>
                        {props.hasPrev && 
                            <MediaSliderButton direction="bottomLeft" onClick={() => {props.onPrevFile(); setIsFlaggedForReset(true)}} />
                        }
                        {props.hasNext &&
                            <MediaSliderButton direction="bottomRight" onClick={() => {props.onNextFile(); setIsFlaggedForReset(true)}} />
                        }
                    </CropperWrapper>
                </CropContainer>
            </ModalSectionWrapper>
            <ModalSectionWrapper style={{width: "100%"}}>
                <Flex>
                    <Slider
                        value={zoom}
                        min={0}
                        max={4}
                        step={0.1}
                        label="Zoom"
                        aria-labelledby="Zoom"
                        onChange={(e: React.ChangeEvent<HTMLInputElement>) => {
                            setZoom(Number.parseFloat(e.target.value))
                        }}
                    />
                    <Slider
                        value={rotation}
                        min={0}
                        max={360}
                        step={1}
                        label="Rotation"
                        aria-labelledby="Rotation"
                        onChange={(e: React.ChangeEvent<HTMLInputElement>) => setRotation(Number.parseFloat(e.target.value))}
                    />                    
                </Flex>
            </ModalSectionWrapper>
        </>
    )
}

export default CreatePostModalCrop;