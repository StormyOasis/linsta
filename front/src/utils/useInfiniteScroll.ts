import React, { useEffect, useRef } from 'react';
import useThrottle from './useThrottle';

/**
 * Type for the async load more function.
 */
type AsyncLoadMoreFunction = () => Promise<void>;

/**
 * Custom hook for infinite scrolling.
 * Attaches a throttled scroll event listener to a container and triggers `loadMore`
 * when the user scrolls near the bottom.
 *
 * @param loadMore - Async function to load more items.
 * @param isLoading - Whether a load is currently in progress.
 * @param containerRef - Ref to the scrollable container element.
 */
const useInfiniteScroll = (loadMore: AsyncLoadMoreFunction, 
    isLoading: boolean, 
    childRef: React.MutableRefObject<HTMLDivElement | null>) => {
        
    const hasScrolled = useRef<boolean>(false);  // Track if the user has scrolled

    // Throttled scroll handler
    const throttledHandleScroll = useThrottle(async () => {
        if (typeof window !== 'undefined' && childRef.current) {
            const element = childRef.current as HTMLElement;
            const currentScroll = window.innerHeight + element.scrollTop;

            if (currentScroll + 256 >= element.scrollHeight && !isLoading && !hasScrolled.current) {
                hasScrolled.current = true;  // Mark as scrolled at least once
                await loadMore();
            }
        }
    }, 200);

    useEffect(() => {
        const childElement = childRef.current;
        if (childElement) {
            childElement.addEventListener('scroll', throttledHandleScroll);
        }

        // Clean up event listener on unmount
        return () => {
            if (childElement) {
                childElement.removeEventListener('scroll', throttledHandleScroll);
            }
        };
    }, [throttledHandleScroll, childRef]);

    // Reset trigger flag when loading finishes
    useEffect(() => {
        if (!isLoading) {
            hasScrolled.current = false;
        }
    }, [isLoading]);
}

export default useInfiniteScroll;