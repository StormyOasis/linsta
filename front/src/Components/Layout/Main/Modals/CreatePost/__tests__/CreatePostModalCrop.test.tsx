// Mock styled-components and react-redux
jest.mock("styled-components", () => {
    const React = require("react");
    function fakeStyled(BaseComponent = "div") {
        const Styled = React.forwardRef((props, ref) =>
            React.createElement(BaseComponent, { ...props, ref }, props.children)
        );
        function templateFn(...args) { return Styled; }
        templateFn.withConfig = () => templateFn;
        Styled.withConfig = () => templateFn;
        Object.setPrototypeOf(templateFn, Styled);
        return templateFn;
    }
    const styled = new Proxy(fakeStyled, {
        get: (target, prop) => fakeStyled(prop),
        apply: (target, thisArg, argArray) => fakeStyled(argArray[0])
    });
    return { __esModule: true, styled, default: styled };
});
jest.mock("../../../../../Common/CombinedStyling", () => ({
    Div: (props: any) => <div {...props} />,
    Flex: (props: any) => <div {...props} />,
}));

jest.mock("react-redux", () => {
    const actual = jest.requireActual("react-redux");
    return {
        ...actual,
        useSelector: jest.fn(),
        useDispatch: () => jest.fn(),
    };
});

jest.mock("../../../../../../utils/utils", () => ({
    isVideoFileFromType: jest.fn(() => false),
}));

jest.mock("../../../../../Common/MultiStepModal", () => ({
    ModalSectionWrapper: (props: any) => <div data-testid="mock-modal-section-wrapper" {...props} />,
}));

jest.mock("../../../../../Common/Slider", () => ({
    __esModule: true,
    default: (props: any) => (
        <input
            data-testid={`mock-slider-${props.label}`}
            type="range"
            value={props.value}
            min={props.min}
            max={props.max}
            step={props.step}
            onChange={props.onChange}
        />
    ),
}));

jest.mock("../../../../../Common/MediaSliderButton", () => ({
    __esModule: true,
    default: (props: any) => (
        <button data-testid={`mock-media-slider-btn-${props.direction}`} onClick={props.onClick} />
    ),
}));

jest.mock("../../../../../Common/Icon", () => ({
    CropSVG: () => <svg data-testid="mock-crop-svg" />,
    FourToFiveSVG: () => <svg data-testid="mock-4-5-svg" />,
    ImageSVG: () => <svg data-testid="mock-image-svg" />,
    OneToOneSVG: () => <svg data-testid="mock-1-1-svg" />,
    SixteenToNineSVG: () => <svg data-testid="mock-16-9-svg" />,
}));

jest.mock("../../../../../Themes/Theme", () => {
    const React = require("react");
    const MockTheme = ({ children }: any) => {
        return <div data-testid="mock-theme">{children}</div>;
    };
    return {
        __esModule: true,
        default: MockTheme
    };
});

jest.mock('react-easy-crop', () => {
    const React = require("react");
    return {
        __esModule: true,
        default: (props: any) => {
            React.useEffect(() => {
                // Simulate what the real Cropper would do
                if (props.onCropComplete) {
                    props.onCropComplete({ x: 0, y: 0, width: 100, height: 100 }, { x: 0, y: 0, width: 100, height: 100 });
                }
            }, []);
            return <div data-testid="mock-cropper" />;
        }
    };
});

import React from "react";

import { screen, fireEvent } from "@testing-library/react";
import CreatePostModalCrop, { defaultCropData } from "../CreatePostModalCrop";
import { renderWithStore } from "../../../../../../utils/test-utils";
import '@testing-library/jest-dom';


describe("CreatePostModalCrop", () => {
    const baseFile = { blob: "blob", type: "image/png" };
    const baseCropData = defaultCropData(1);

    it("renders nothing if files is null", () => {
        const r = renderWithStore(
            <CreatePostModalCrop files={null as any} cropData={[]} setCropData={jest.fn()} />
        );
        const themeDiv = r.container.querySelector('[data-testid="mock-theme"]');
        expect(themeDiv).toBeInTheDocument();
        expect(themeDiv?.children.length).toBe(0);
    });

    it("renders nothing if files is empty", () => {
        const r = renderWithStore(
            <CreatePostModalCrop files={[]} cropData={[]} setCropData={jest.fn()} />
        );
        const themeDiv = r.container.querySelector('[data-testid="mock-theme"]');
        expect(themeDiv).toBeInTheDocument();
        expect(themeDiv?.children.length).toBe(0); 
    });

    it("renders cropper and sliders for image file", () => {
        const r = renderWithStore(
            <CreatePostModalCrop files={[baseFile]} cropData={baseCropData} setCropData={jest.fn()} />
        );
        expect(screen.getByTestId("mock-cropper")).toBeInTheDocument();
        expect(screen.getByTestId("mock-slider-Zoom")).toBeInTheDocument();
        expect(screen.getByTestId("mock-slider-Rotation")).toBeInTheDocument();
        expect(r.container).toMatchSnapshot();
    });

    it("renders video warning if file is video", () => {
        const { isVideoFileFromType } = require("../../../../../../utils/utils");
        isVideoFileFromType.mockReturnValueOnce(true);
        const r = renderWithStore(
            <CreatePostModalCrop files={[{ ...baseFile, type: "video/mp4" }]} cropData={baseCropData} setCropData={jest.fn()} />
        );
        expect(screen.getByText(/Editing video files is currently unsupported/i)).toBeInTheDocument();
        expect(r.container).toMatchSnapshot();
    });

    it("handles next/prev file navigation", () => {
        const files = [baseFile, baseFile];
        const cropData = defaultCropData(2);
        const r = renderWithStore(
            <CreatePostModalCrop files={files} cropData={cropData} setCropData={jest.fn()} />
        );
        // Next button should be present
        expect(screen.getByTestId("mock-media-slider-btn-bottomRight")).toBeInTheDocument();
        fireEvent.click(screen.getByTestId("mock-media-slider-btn-bottomRight"));
        // Prev button should now be present
        expect(screen.getByTestId("mock-media-slider-btn-bottomLeft")).toBeInTheDocument();
        fireEvent.click(screen.getByTestId("mock-media-slider-btn-bottomLeft"));
        expect(r.container).toMatchSnapshot();
    });

    it("toggles aspect ratio menu", () => {
        const r = renderWithStore(
            <CreatePostModalCrop files={[baseFile]} cropData={baseCropData} setCropData={jest.fn()} />
        );
        const aspectBtn = screen.getByTestId("mock-crop-svg").closest("div");
        fireEvent.click(aspectBtn!);
        // The aspect ratio menu should now be open (rendered)
        expect(r.container).toMatchSnapshot();
    });
});