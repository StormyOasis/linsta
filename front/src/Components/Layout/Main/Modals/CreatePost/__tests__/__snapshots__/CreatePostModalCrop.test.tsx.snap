// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreatePostModalCrop handles next/prev file navigation 1`] = `
<div>
  <div
    data-testid="mock-theme"
  >
    <div
      data-testid="mock-modal-section-wrapper"
    >
      <div>
        <div>
          <div>
            <div
              data-testid="mock-cropper"
            />
          </div>
          <div>
            <div>
              <svg
                data-testid="mock-crop-svg"
              />
            </div>
          </div>
          <button
            data-testid="mock-media-slider-btn-bottomRight"
          />
        </div>
      </div>
    </div>
    <div
      data-testid="mock-modal-section-wrapper"
      style="width: 100%;"
    >
      <div>
        <input
          data-testid="mock-slider-Zoom"
          max="4"
          min="0"
          step="0.1"
          type="range"
          value="1"
        />
        <input
          data-testid="mock-slider-Rotation"
          max="360"
          min="0"
          step="1"
          type="range"
          value="0"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`CreatePostModalCrop renders cropper and sliders for image file 1`] = `
<div>
  <div
    data-testid="mock-theme"
  >
    <div
      data-testid="mock-modal-section-wrapper"
    >
      <div>
        <div>
          <div>
            <div
              data-testid="mock-cropper"
            />
          </div>
          <div>
            <div>
              <svg
                data-testid="mock-crop-svg"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      data-testid="mock-modal-section-wrapper"
      style="width: 100%;"
    >
      <div>
        <input
          data-testid="mock-slider-Zoom"
          max="4"
          min="0"
          step="0.1"
          type="range"
          value="1"
        />
        <input
          data-testid="mock-slider-Rotation"
          max="360"
          min="0"
          step="1"
          type="range"
          value="0"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`CreatePostModalCrop renders video warning if file is video 1`] = `
<div>
  <div
    data-testid="mock-theme"
  >
    <div>
      Note: Editing video files is currently unsupported
    </div>
    <div
      data-testid="mock-modal-section-wrapper"
    >
      <div>
        <div>
          <div>
            <div
              data-testid="mock-cropper"
            />
          </div>
          <div>
            <div>
              <svg
                data-testid="mock-crop-svg"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      data-testid="mock-modal-section-wrapper"
      style="width: 100%;"
    >
      <div>
        <input
          data-testid="mock-slider-Zoom"
          max="4"
          min="0"
          step="0.1"
          type="range"
          value="1"
        />
        <input
          data-testid="mock-slider-Rotation"
          max="360"
          min="0"
          step="1"
          type="range"
          value="0"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`CreatePostModalCrop toggles aspect ratio menu 1`] = `
<div>
  <div
    data-testid="mock-theme"
  >
    <div
      data-testid="mock-modal-section-wrapper"
    >
      <div>
        <div>
          <div>
            <div
              data-testid="mock-cropper"
            />
            <div>
              <div>
                <div>
                  <div>
                    Original
                  </div>
                  <svg
                    data-testid="mock-image-svg"
                  />
                </div>
              </div>
              <div>
                <div>
                  <div>
                    1:1
                  </div>
                  <svg
                    data-testid="mock-1-1-svg"
                  />
                </div>
              </div>
              <div>
                <div>
                  <div>
                    1:2
                  </div>
                </div>
              </div>
              <div>
                <div>
                  <div>
                    4:3
                  </div>
                </div>
              </div>
              <div>
                <div>
                  <div>
                    4:5
                  </div>
                  <svg
                    data-testid="mock-4-5-svg"
                  />
                </div>
              </div>
              <div>
                <div>
                  <div>
                    16:9
                  </div>
                  <svg
                    data-testid="mock-16-9-svg"
                  />
                </div>
              </div>
            </div>
          </div>
          <div>
            <div>
              <svg
                data-testid="mock-crop-svg"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      data-testid="mock-modal-section-wrapper"
      style="width: 100%;"
    >
      <div>
        <input
          data-testid="mock-slider-Zoom"
          max="4"
          min="0"
          step="0.1"
          type="range"
          value="1"
        />
        <input
          data-testid="mock-slider-Rotation"
          max="360"
          min="0"
          step="1"
          type="range"
          value="0"
        />
      </div>
    </div>
  </div>
</div>
`;
